examples/gaussian_process/svgp.py

```diff
@@ -0,0 +1,216 @@
+    x_train, x_test, _, _ = dataset.standardize(x_train, x_test)
+    y_train, y_test, mean_y_train, std_y_train = dataset.standardize(
+        y_train, y_test)
+
+    # Build graph
+    X_ph = tf.placeholder(hps.dtype, [None, hps.N_X], 'X')
+    Y_ph = tf.placeholder(hps.dtype, [None], 'Y')
+    Z_pos = tf.get_variable('Z/pos', [hps.N_Z, hps.N_X], hps.dtype,
+                            initializer=tf.random_uniform_initializer(-1, 1))
+    N_particles = tf.placeholder(tf.int32, [], 'n_particles')
+
+    # Train
+    def log_joint(observed):
+        model, _ = build_model(observed, Z_pos, X_ph, hps, N_particles)
+        prior, ll = model.local_log_prob(['fZ', 'Y'])
+        n_batch = tf.cast(tf.shape(X_ph)[0], hps.dtype)
+        return prior + ll / n_batch * N
+
+    variational = build_variational(hps, N_particles)
+    var_fZ = variational.query('fZ', outputs=True, local_log_prob=True)
+    lower_bound = zs.variational.elbo(
+        log_joint, observed={'Y': Y_ph}, latent={'fZ': var_fZ}, axis=0)
+    cost = tf.reduce_mean(lower_bound.sgvb())
+    lower_bound = tf.reduce_mean(lower_bound)
+    optimizer = tf.train.AdamOptimizer(learning_rate=0.01)
+    infer_op = optimizer.minimize(cost)
+
+    def infer_step(X_batch, Y_batch):
+        fd = {X_ph: X_batch, Y_ph: Y_batch, N_particles: hps.N_particles}
+        return sess.run([infer_op, cost], fd)[1]
+
+    # Predict
+    observed = {'fZ': var_fZ[0], 'Y': Y_ph}
+    model, y_inferred = build_model(observed, Z_pos, X_ph, hps, N_particles)
+    lhood = model.local_log_prob('Y')
+    N_Y = tf.cast(tf.shape(Y_ph)[0], hps.dtype)
+
+    def log_mean_exp(x, axis=None, keep_dims=False):
+        # FIXME: Any reason zs.log_mean_exp cast to tf.float32?
+        x_max = tf.reduce_max(x, axis=axis, keep_dims=True)
+        ret = tf.log(tf.reduce_mean(tf.exp(x - x_max), axis=axis,
+                                    keep_dims=True)) + x_max
+        if not keep_dims:
+            ret = tf.reduce_mean(ret, axis=axis)
+        return ret
+
+    std_y_train = tf.cast(std_y_train, hps.dtype)
+    lhood = log_mean_exp(lhood, 0) / N_Y - tf.log(std_y_train)
+    y_pred_mean = tf.reduce_mean(y_inferred.distribution.mean, axis=0)
+    pred_mse = tf.reduce_mean((y_pred_mean - Y_ph) ** 2)
+    pred_mse *= std_y_train ** 2
+
+    def predict_step(X_batch, Y_batch):
+        fd = {X_ph: X_batch, Y_ph: Y_batch, N_particles: hps.N_particles_test}
+        return sess.run([lhood, pred_mse], fd)
+
+    # Run the inference
+    iters = int(np.ceil(x_train.shape[0] / float(hps.N_batch)))
+    test_freq = 10
+    with tf.Session() as sess:
+        sess.run(tf.global_variables_initializer())
+        for epoch in range(1, hps.N_epoch + 1):
+            lbs = []
+            indices = np.arange(x_train.shape[0])
+            np.random.shuffle(indices)
+            for t in range(iters):
+                idx = indices[t * hps.N_batch: (t + 1) * hps.N_batch]
+                lb = infer_step(x_train[idx], y_train[idx])
```

Comment:
Should do `x_train[idx]` and `y_train[idx]` outside of the inner loop.

Reply:
Did you mean that we should write `infer_step(indices)` instead, or could you clarify?

Reply:
No, I mean in each iteration a new copy of the `x_train` and `y_train` is created by
`x_train[idx]` and `y_train[idx]`, which is only needed once per epoch.

Reply:
Fixed.